File: typescript/src/core/modules/auth/helpers/get-public-keys.ts
import axios from "axios";

// Import constants
import { Configs } from "../../../../utils/configs";

// Import errors
import { AppError } from "../../../error";

const region = Configs.AWSRegion;
const userPoolId = process.env.COGNITO_USER_POOL_ID;
const jwksURL = `https://cognito-idp.${region}.amazonaws.com/${userPoolId}/.well-known/jwks.json`;

/**
 * Lấy các public keys trong cognito user pool.
 *
 * @returns
 */
export async function getPublicKeys(): Promise<Array<any> | AppError> {
  try {
    const response = await axios.get(jwksURL);
    const data = response.data;
    return data.keys as Array<any>;
  } catch (error: any) {
    const err = new AppError("Cannot get public keys from Cognito User Pool");
    err.asHTTPError("InternalServerError");
    err.addErrorDetail({ source: "getPublicKeys", desc: error.message });
    return err;
  }
}

File: python/src/core/modules/auth/helpers/get_public_keys.py
import os
import requests

from core.error.AppError import AppError

# Import constants
from utils.configs.main import Configs


def get_public_keys():
    """
    Lấy các public keys trong Cognito User Pool.
    Trả về list keys hoặc AppError nếu thất bại.
    """
    region = Configs.Aws_Region
    user_pool_id = Configs.Cognito_User_Pool_Id

    jwks_url = f"https://cognito-idp.{region}.amazonaws.com/{user_pool_id}/.well-known/jwks.json"

    try:
        response = requests.get(jwks_url, timeout=10)
        response.raise_for_status()
        data = response.json()

        return data.get("keys", [])
    except requests.RequestException as e:
        err = AppError("Cannot get public keys from Cognito User Pool")
        err.add_error_detail({"source": "get_public_keys", "desc": str(e)})
        err.as_http_error("InternalServerError")

        return err
